### PR TITLE
Generate Definitions

### LINKED ISSUE #101: Gateway timeouts when computing codelist definitions
Codelist definitions are computed on the fly, but this is slow for large SNOMED CT codelists, and can lead to gateway timeouts.

One solution might be to turn on caching, in Cloudflare or elsewhere.

Another solution might be to precompute definitions when a codelist CSV is uploaded.  For now, the definition can be stored as JSON in the database.  As it is only internal users who upload CSVs, we wouldn't need to add a job queue to handle this.

### PR DESCRIPTION
**Pull request by ghickman**
This adds a function to generate Definitions as a list of dictionaries, so we can store them as a JSON structure on a given CodelistVersion.  The generation method (`build_definition`) mirrors the structure of `build_html_definition` but creates a data structure holding only the data necessary for rendering.  Styling is left up to the template, and URL generation to the included templatetag.

Definitions are constructed when a Version is created or updated.  There is also a management command included to aid local testing and backfilling in Production.

Fixes #101

---

I've made use of Attrs and types for the Row class.  I think this pattern shook out well for this small use case.  The definition is clear and readable, there's no inheritance soup, it gave me a place to clearly document what the various fields are for, and it gets exported to a dictionary (ready for JSON dumping into the DB) so the class doesn't leak out of `iter_definitions`.  I'm only using the types to aid readability here (there's nothing enforcing those types) and I've not even added that `excluded_children` is really `List[Row]`.  I found small pointers like this were useful while I was getting my head around converting the previous implementation.  What do you think?

One part I'm unsure of is `_build_definition` in `actions.py`.  It doesn't really feel like an action but I'm not sure where else it should go given we do want to wrap up building defintions.  Putting it in `definition.py` looks wrong to me because that part is tied to the model layer.

cc @evansd & @sebbacon - especially for thoughts on attrs/types!

### CODE REVIEW

File: codelists/actions.py [hunk 1]
```diff
@@ -91,6 +115,9 @@ def update_version(*, version, csv_data):
     version.csv_data = csv_data
     version.save()
 
+    version.definition = _build_definition(version)
+    version.save()
```
Comment (inglesp) [inline codelists/actions.py:119]: Transaction here too.

**Comment by sebbacon** [comment]
I like the `attrs` pattern - definitely nicer than named tuples. I'm not 100% convinced the extra overhead of expecting someone to know how `attrs` works is worth the saving of writing a couple of dunder methods... but I'm 90% convinced :)

File: codelists/templatetags/definitions.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+from django import template
+from django.urls import reverse
+
+register = template.Library()
+
+
+@register.simple_tag
+def definition_url(coding_system_id, code):
+    return reverse(f"{coding_system_id}:concept", args=[code])
```
[the remark this answers is not in the document]
Comment (ghickman) [reply codelists/templatetags/definitions.py:9]: It's right there in the URL name and everything 🤦‍♂️ thanks!

File: codelists/views.py [hunk 3]
```diff
@@ -12,7 +12,7 @@
 
 from . import actions, tree_utils
 from .coding_systems import CODING_SYSTEMS
-from .definition import Definition, build_html_definition
```
[the remark this answers is not in the document]
Comment (ghickman) [reply codelists/views.py:15]: You asked me not to! 😂

File: codelists/definition.py [hunk 4]
```diff
@@ -149,6 +150,70 @@ def negative_helper(tree):
...
+        * all descendants are included (all_descendants = True, excluded_children = [])
+        * all descendants except N (all_descendants = True, excluded_children = [...])
+        * no descendants (all_descendants = False, excluded_children ignored)
+
+    """
+
+    name: str = attr.ib()
+    code: str = attr.ib()
+    excluded_children: list = attr.ib(default=list())
+    all_descendants: bool = attr.ib(default=True)
```
[the remark this answers is not in the document]
Comment (ghickman) [reply codelists/definition.py:175]: I _think_ it's "just" applying mypy to the code but I've not played with it at all, shall I have a go with it after this PR lands?

File: codelists/views.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (inglesp) [reply codelists/views.py:15]: Did I?  I've changed what I please to call my mind.  Unless you can remember why I said that, delete it now before I change it back!!

File: codelists/definition.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (inglesp) [reply codelists/definition.py:175]: Should think it's worth spending an hour on.

File: codelists/actions.py (hunk 1 above)
Comment (ghickman) [reply codelists/actions.py:104]: Fixed in 9c8ad19

File: codelists/actions.py [hunk 5]
```diff
@@ -63,8 +78,17 @@ def create_version(*, codelist, csv_data):
...
-    raise ValueError("E_TOO_MANY_VERSIONS")
+            version = codelist.versions.create(
+                version_str=version_str, csv_data=csv_data
+            )
+            break
+    else:
+        raise ValueError("E_TOO_MANY_VERSIONS")
+
+    version.definition = _build_definition(version)
+    version.save()
```
[the remark this answers is not in the document]
Comment (ghickman) [reply codelists/actions.py:75]: Fixed the transaction part in 9c8ad19

File: codelists/models.py [hunk 6]
```diff
@@ -54,6 +54,8 @@ class CodelistVersion(models.Model):
     csv_data = models.TextField(verbose_name="CSV data")
     is_draft = models.BooleanField(default=True)
 
+    definition = models.JSONField(null=True)
```
[the remark this answers is not in the document]
Comment (ghickman) [reply codelists/models.py]: Fixed in 8c91d6d

File: codelists/actions.py [hunk 7]
```diff
@@ -5,6 +5,21 @@
 
 from opencodelists.models import User
 
+from . import tree_utils
+from .coding_systems import CODING_SYSTEMS
+from .definition import Definition, build_definition
+
+
+def _build_definition(version):
```
[the remark this answers is not in the document]
Comment (ghickman) [reply codelists/actions.py]: Fixed in 1ceccd3

File: codelists/views.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ghickman) [reply codelists/views.py:15]: Fixed in 69cd901

File: codelists/templatetags/definitions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ghickman) [reply codelists/templatetags/definitions.py:9]: Fixed in fdb6cdf

File: codelists/definition.py [hunk 8]
```diff
@@ -1,3 +1,4 @@
+import attr
```
[the remark this answers is not in the document]
Comment (ghickman) [reply codelists/definition.py:1]: Fixed in 234c37e

**Comment by ghickman** [comment]
Closing in favour of #149 which rebuilds these changes on top of @inglesp's new `Hierarchy` object and uses the presenters pattern we've switch to (which nicely groups this with other similar functions).

**Pull request closed without merging**